### PR TITLE
Refactor SpearmanRankCorrelation to native PyTorch and remove SciPy dependency

### PR DESCRIPTION
**Pull request by Prathamesh8989**
Fixes #3663

## Description
This PR refactors the `SpearmanRankCorrelation` metric to use native PyTorch operations, effectively removing the hard dependency on `scipy` for calculation.

### Key Changes:
- **Native Ranking Implementation:** Added a `_get_ranks` helper function that calculates ranks using the "average" method for ties, ensuring parity with `scipy.stats.spearmanr`.
- **Logic Update:** Replaced the SciPy-based `_spearman_r` function with a pure PyTorch implementation using Pearson correlation on ranked tensors.
- **Robust Testing:** Updated the test suite to treat SciPy as an optional "gold standard" check. If SciPy is missing, tests skip gracefully instead of crashing, but the metric remains fully functional.
- **Improved Performance:** By keeping operations in PyTorch, we reduce unnecessary CPU/GPU synchronization and data transfers.

### Check list:
- [x] New tests are added (Updated existing tests to handle optional dependency)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (Updated "Implementation" section in docstrings)

### CODE REVIEW

File: ignite/metrics/regression/spearman_correlation.py [hunk 1]
```diff
@@ -10,13 +10,55 @@
...
+    mu_y = torch.mean(r_targets)
+
+    diff_x = r_preds - mu_x
+    diff_y = r_targets - mu_y
+
+    norm_x = torch.norm(diff_x, 2)
+    norm_y = torch.norm(diff_y, 2)
+
+    if norm_x == 0 or norm_y == 0:
+        return 0.0
```
Comment (TahaZahid05) [inline ignite/metrics/regression/spearman_correlation.py]: Scipy returns Nan in such cases?

File: ignite/metrics/regression/spearman_correlation.py [hunk 2]
```diff
@@ -10,13 +10,55 @@
...
+    # Use cumsum of repetitions to find the range of ranks for each unique value
+    right = torch.cumsum(repetitions, dim=0)
+    left = right - repetitions + 1
+    avg_ranks = (left + right).float() / 2.0
+
+    # Map back to original order
+    return avg_ranks[dense_ranks - 1][inv_sorter]
+
+
 def _spearman_r(predictions: Tensor, targets: Tensor) -> float:
```
Comment (TahaZahid05) [inline ignite/metrics/regression/spearman_correlation.py:39]: There are no checks for NaN

File: tests/ignite/metrics/regression/test_spearman_correlation.py [hunk 3]
```diff
@@ -139,7 +151,7 @@ def test_compute(self):
             assert pytest.approx(np_ans, rel=2e-4) == m.compute()
 
     @pytest.mark.parametrize("n_epochs", [1, 2])
-    def test_integration(self, n_epochs: int):
+    def test_integration(self, n_epochs: int):  # <--- FIXED: changed from self_n_epochs to n_epochs
```
Comment (TahaZahid05) [inline tests/ignite/metrics/regression/test_spearman_correlation.py]: Why is there a need to rename here?

File: tests/ignite/metrics/regression/test_spearman_correlation.py [hunk 4]
```diff
@@ -167,7 +179,7 @@ def test_integration(self, n_epochs: int):
             corr.attach(engine, "spearman_corr")
 
             data = list(range(n_iters))
-            engine.run(data=data, max_epochs=n_epochs)
+            engine.run(data=data, max_epochs=n_epochs)  # <--- FIXED: using the parameter correctly now
```
Comment (TahaZahid05) [inline tests/ignite/metrics/regression/test_spearman_correlation.py]: Why did you add this comment? The line is the same?

File: tests/ignite/metrics/regression/test_spearman_correlation.py [hunk 5]
```diff
@@ -1,14 +1,24 @@
...
+# Handle optional scipy for testing purposes (the "gold standard" comparison)
+try:
+    from scipy.stats import spearmanr
+
+    HAS_SCIPY = True
+except ImportError:
+    HAS_SCIPY = False
+
+# Skip the tests that require SciPy comparison if it's not installed
+pytestmark = pytest.mark.skipif(not HAS_SCIPY, reason="SciPy is required for comparison tests")
```
Comment (TahaZahid05) [inline tests/ignite/metrics/regression/test_spearman_correlation.py]: Not the correct way to handle this. Tests shouldn't be skipped. User should be expected to follow `requirement-dev`

**Review by TahaZahid05 — CHANGES_REQUESTED** [review]

**Comment by Prathamesh8989** [comment]
Hi @TahaZahid05,

Thank you for the feedback!

I've addressed the suggested changes and updated the implementation accordingly. Please let me know if there are any further adjustments needed from my side.

File: ignite/metrics/regression/spearman_correlation.py [hunk 6]
```diff
@@ -10,13 +10,58 @@
...
+    dense_ranks = torch.cumsum(obs, dim=0)
+
+    # Calculate average ranks for ties
+    count = torch.cat([torch.nonzero(obs).flatten(), torch.tensor([n], device=x.device)])
+    repetitions = count[1:] - count[:-1]
+
+    # Use cumsum of repetitions to find the range of ranks for each unique value
+    right = torch.cumsum(repetitions, dim=0)
+    left = right - repetitions + 1
+    avg_ranks = (left + right).float() / 2.0
```
Comment (TahaZahid05) [inline ignite/metrics/regression/spearman_correlation.py]: `double()` is more safe i think

File: tests/ignite/metrics/regression/test_spearman_correlation.py
Comment (TahaZahid05) [inline tests/ignite/metrics/regression/test_spearman_correlation.py:1]: - Add test so metric returns `NaN` when input tensor contains `NaN` values.
- Add test for constant inputs as well
- Verify average rank calculation by adding a test with intended values such as `[10, 20, 20, 30]` to test the logic.

**Review by TahaZahid05 — CHANGES_REQUESTED** [review]

**Comment by Prathamesh8989** [comment]
Hi @TahaZahid05,

Thanks for the suggestions!

I've made the requested updates:

Switched the rank calculation to use `.double()` for improved numerical stability.
Added tests to ensure the metric returns NaN when inputs contain NaN values.
Added a test for constant inputs.
Added a test verifying the average rank calculation using `[10, 20, 20, 30]`.

Please let me know if any further adjustments are needed.

File: ignite/metrics/regression/spearman_correlation.py
Comment (TahaZahid05) [inline ignite/metrics/regression/spearman_correlation.py:122]: Add versionchanged to 0.5.5 and state the change

**Review by TahaZahid05 — APPROVED** [review]
Just one last change, else everything else is good!

**Comment by Prathamesh8989** [comment]
@TahaZahid05, 
Thanks for the suggestion! I've added the `versionchanged:: 0.5.5` entry in the docstring to document the switch to the native PyTorch implementation and removal of the SciPy dependency.

Please let me know if anything else needs adjustment.

**Review by TahaZahid05 — APPROVED** [review]
Thanks for the changes @Prathamesh8989. @vfdev-5 Looks good to me!

**Comment by TahaZahid05** [comment]
@Prathamesh8989, style checks are failing. Kindly fix them.

**Comment by Prathamesh8989** [comment]
yeah sure , fixing them.

**Review by TahaZahid05 — APPROVED** [review]

File: requirements-dev.txt [hunk 7]
```diff
@@ -1,6 +1,7 @@
 # Tests
 dill
 filelock
+packaging
```
Comment (vfdev-5) [inline requirements-dev.txt]: Why do we add packaging here?
We already have it as ignite depedency here: 
https://github.com/pytorch/ignite/blob/6c3dfd9c2d66815d35f4a2b49047f27609ebcf35/pyproject.toml#L20

Comment (Prathamesh8989) [reply requirements-dev.txt]: it gave me an error saying that the tests failed because of not having this , so tried to add this in requirements.txt

Comment (vfdev-5) [reply requirements-dev.txt]: Let's remove it

**Comment by TahaZahid05** [comment]
@Prathamesh8989 can you please verify your pre-commit installation? The code style checks are still failing. You may take a look at `CONTRIBUTING.MD` for more reference.

**Comment by Prathamesh8989** [comment]
sure, i will get through it by tomorrow.

**Comment by TahaZahid05** [comment]
@Prathamesh8989 it seems you have pushed unwanted changes when fixing the code style checks. All files in examples are also coming under file changed. Can you fix it?

**Comment by Prathamesh8989** [comment]
Hey @TahaZahid05,

I ran pre-commit run --all-files, then staged, committed, and pushed the files. I believe that should be the correct workflow, but I might be missing something. Could you please guide me a bit on this? I haven’t yet figured out the proper way to verify it.

Also, I’ll be having my practical and theory exams from 22nd to 30th April, so my response time might be a bit slow during that period. Kindly excuse the delay in advance.

**Review by TahaZahid05 — APPROVED** [review]

**Comment by TahaZahid05** [comment]
@Prathamesh8989 Run the following set of commands:

```bash
pip install -U pre-commit 
pre-commit clean           
pre-commit install         
pre-commit run --all-files 
```

You'll see in your terminal:

```bash
nb-clean.................................................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
prettier.................................................................Passed
ruff check...............................................................Passed
ruff format..............................................................Failed
- hook id: ruff-format
- files were modified by this hook

1 file reformatted, 364 files left unchanged
```

Then run `git add .` again to stage the change pre-commit made, and then commit and push. Follow along these steps and let me know where your output is deviating or if you get stuck somewhere, I can help you out from there. You can contact me on Discord if you need further help

**Comment by vfdev-5** [comment]
@Prathamesh8989 the failure on MPS is real: https://github.com/pytorch/ignite/actions/runs/24736092143/job/72390524588?pr=3672
MPS does not have f64.

**Comment by TahaZahid05** [comment]
@Prathamesh8989 hi! are you still working on this?
